### PR TITLE
Vehicle Racks

### PR DESCRIPTION
**Pull request by Sniperhid**
**When merged this pull request will:**
- Update sys_components module - fix broken function, document headers and add some more useful funcitons. Lots of ground work done for allowing unique components to be vehicles instead of items.
- Update ace_interact module - Vehicle rack interaction is all done through ace interaction. Also adds support for exploring the connectors but this is disabled by default.
- Add new module sys_rack - Handles the backend and data side for vehicle racks
- Adds the racks AN/VRC110 for 152s and AN/VRC103 for 117F radios

### CODE REVIEW

File: addons/ace_interact/fnc_initVehicle.sqf [hunk 1]
```diff
@@ -25,6 +25,7 @@ if (_type in GVAR(initializedVehicleClasses)) exitWith {};
 GVAR(initializedVehicleClasses) pushBack _type;
 
 if (!hasInterface) exitWith {};
+if (isNil "ace_interact_menu_fnc_addActionToClass") exitWith {} // No ACE exit.
```
Comment (jonpas) [inline addons/ace_interact/fnc_initVehicle.sqf]: This should be changed to `!isClass (configFile >> "CfgPatches" >> "ace_interact_menu")`

File: addons/sys_prc117f/radio/fnc_getCurrentChannelData.sqf [hunk 2]
```diff
@@ -53,7 +53,11 @@ switch _channelType do {
         HASH_SET(_return, "mode", "singleChannel");
         HASH_SET(_return, "frequencyTX", HASH_GET(_currentChannelData, "frequencyTX"));
         HASH_SET(_return, "frequencyRX", HASH_GET(_currentChannelData, "frequencyRX"));
-        HASH_SET(_return, "power", HASH_GET(_currentChannelData, "power"));
+        if (HASH_GET(_radioData, "pgm_pa_mode") == "ON" && {HASH_GET(_radioData, "powerSource") == "VAU"}) then {
+            HASH_SET(_return, "power", 50000);
+        } else {
+            HASH_SET(_return, "power", HASH_GET(_channel, "power"));
```
Comment (Soldia1138) [inline addons/sys_prc117f/radio/fnc_getCurrentChannelData.sqf]: `HASH_GET(_channel, "power")` -> `HASH_GET(_currentChannelData, "power")`

File: addons/sys_prc148/radio/fnc_getChannelDataInternal.sqf [hunk 3]
```diff
@@ -28,7 +28,11 @@ switch _channelType do {
         HASH_SET(_return, "mode", "singleChannel");
         HASH_SET(_return, "frequencyTX", HASH_GET(_channel, "frequencyTX"));
         HASH_SET(_return, "frequencyRX", HASH_GET(_channel, "frequencyRX"));
-        HASH_SET(_return, "power", HASH_GET(_channel, "power"));
+        if (_powerSource == "BAT") then {
```
Comment (Soldia1138) [inline addons/sys_prc148/radio/fnc_getChannelDataInternal.sqf:33]: `_powerSource` not defined

File: addons/sys_prc152/radio/fnc_getCurrentChannelData.sqf [hunk 4]
```diff
@@ -47,13 +47,18 @@ if (!(isNil "_optChannelId") && !(isNil "_opt")) then {
...
+private _powerSource = HASH_GET(_radioData, "powerSource");
 private _return = HASH_CREATE;
 switch _channelType do {
     case "BASIC": {
         HASH_SET(_return, "mode", "singleChannel");
         HASH_SET(_return, "frequencyTX", HASH_GET(_currentChannelData, "frequencyTX"));
         HASH_SET(_return, "frequencyRX", HASH_GET(_currentChannelData, "frequencyRX"));
-        HASH_SET(_return, "power", HASH_GET(_currentChannelData, "power"));
+        if (_powerSource == "BAT") then {
+            HASH_SET(_return, "power", HASH_GET(_channel, "power"));
```
Comment (Soldia1138) [inline addons/sys_prc152/radio/fnc_getCurrentChannelData.sqf]: `HASH_GET(_channel, "power")` -> `HASH_GET(_currentChannelData, "power")`

File: addons/sys_prc148/radio/fnc_getChannelDataInternal.sqf [hunk 5]
```diff
@@ -19,16 +19,22 @@
...
 switch _channelType do {
     case "BASIC": {
         HASH_SET(_return, "mode", "singleChannel");
         HASH_SET(_return, "frequencyTX", HASH_GET(_channel, "frequencyTX"));
         HASH_SET(_return, "frequencyRX", HASH_GET(_channel, "frequencyRX"));
-        HASH_SET(_return, "power", HASH_GET(_channel, "power"));
+        if (_powerSource == "BAT") then {
+            HASH_SET(_return, "power", HASH_GET(_channel, "power"));
+        } else { // RACK BOOST
+            HASH_SET(_return, "power", 20000);
```
Comment (Soldia1138) [inline addons/sys_prc148/radio/fnc_getChannelDataInternal.sqf]: Again, magic number -> macro or GVAR

File: addons/sys_prc152/radio/fnc_getChannelData.sqf [hunk 6]
```diff
@@ -29,7 +30,11 @@ switch _channelType do {
         HASH_SET(_return, "mode", "singleChannel");
         HASH_SET(_return, "frequencyTX", HASH_GET(_channel, "frequencyTX"));
         HASH_SET(_return, "frequencyRX", HASH_GET(_channel, "frequencyRX"));
-        HASH_SET(_return, "power", HASH_GET(_channel, "power"));
+        if (_powerSource == "BAT") then {
+            HASH_SET(_return, "power", HASH_GET(_channel, "power"));
+        } else { // RACK BOOST
+            HASH_SET(_return, "power", 20000);
```
Comment (Soldia1138) [inline addons/sys_prc152/radio/fnc_getChannelData.sqf]: Again, magic number -> macro or GVAR

File: addons/sys_prc152/radio/fnc_getCurrentChannelData.sqf [hunk 7]
```diff
@@ -47,13 +47,18 @@ if (!(isNil "_optChannelId") && !(isNil "_opt")) then {
...
 switch _channelType do {
     case "BASIC": {
         HASH_SET(_return, "mode", "singleChannel");
         HASH_SET(_return, "frequencyTX", HASH_GET(_currentChannelData, "frequencyTX"));
         HASH_SET(_return, "frequencyRX", HASH_GET(_currentChannelData, "frequencyRX"));
-        HASH_SET(_return, "power", HASH_GET(_currentChannelData, "power"));
+        if (_powerSource == "BAT") then {
+            HASH_SET(_return, "power", HASH_GET(_currentChannelData, "power"));
+        } else { // RACK BOOST
+            HASH_SET(_return, "power", 20000);
```
Comment (Soldia1138) [inline addons/sys_prc152/radio/fnc_getCurrentChannelData.sqf]: Again, magic number -> macro or GVAR

File: addons/sys_rack/$PBOPREFIX$ [hunk 8]
```diff
@@ -0,0 +1 @@
+idi\acre\addons\sys_rack
```
Comment (Soldia1138) [inline addons/sys_rack/$PBOPREFIX$]: No New Line at EOF

File: addons/sys_rack/fnc_addRackOnReturn.sqf [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+#include "script_component.hpp"
+
+params ["_vehicle","_rackId"];
+
+private _rack = _rackId createVehicle [0,0,0];
+if (isNull _rack) exitWith {
+    WARNING_2("Error creating rack '%2' for vehicle %1", str _vehicle, _rackId);
+};
+
+_rack attachTo [_vehicle,[0,0,0]];
```
Comment (Soldia1138) [inline addons/sys_rack/fnc_addRackOnReturn.sqf:26]: As discussed in slack...we should test network performance on `attachTo`

File: addons/sys_rack/fnc_configWiredIntercoms.sqf [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+if (count _accessibleIntercoms ==  0) then {
+    if (_hasCrewIntercom == 1) then {
+        _accessibleIntercoms = ["crew"];
+    };
+
+    if ((_hasPassengerIntercom == 1) && count (_vehicle getVariable [QGVAR(crewIntercomPositions), []]) > 0) then {
+        _accessibleIntercoms pushBack "passenger";
+    };
+} else {
+
```
Comment (Soldia1138) [inline addons/sys_rack/fnc_configWiredIntercoms.sqf]: Remove blank line

File: addons/sys_rack/fnc_onReturnRackId.sqf [hunk 11]
```diff
@@ -0,0 +1,57 @@
...
+LOG("HIT CALLBACK");
+
+params ["_entity", "_class", "_returnIdNumber", "_replacementId"];
+
+private _dataHash = HASH_CREATE;
+
+// diag_log text format ["acre_sys_data_radioData: %1", acre_sys_data_radioData];
+
+HASH_SET(EGVAR(sys_data,radioData), _class, _dataHash);
+_idRelation = [_entity, _entity];
```
Comment (Soldia1138) [inline addons/sys_rack/fnc_onReturnRackId.sqf]: missing `private`

File: addons/sys_rack/fnc_onReturnRackId.sqf [hunk 12]
```diff
@@ -0,0 +1,57 @@
...
+HASH_SET(EGVAR(sys_server,objectIdRelationTable), _class, _idRelation);
+if (_replacementId != "") then {
+    _radioData = HASH_GET(EGVAR(sys_data,radioData), _replacementId);
```
Comment (Soldia1138) [inline addons/sys_rack/fnc_onReturnRackId.sqf]: missing `private`

File: addons/sys_rack/fnc_onReturnRadioId.sqf [hunk 13]
```diff
@@ -0,0 +1,61 @@
...
+ * Public: No
+ */
+#include "script_component.hpp"
+
+params ["_rackObject", "_class", "_returnIdNumber"];
+
+private _dataHash = HASH_CREATE;
+
+HASH_SET(EGVAR(sys_data,radioData),_class,_dataHash);
+_idRelation = [_rackObject, _rackObject];
```
Comment (Soldia1138) [inline addons/sys_rack/fnc_onReturnRadioId.sqf]: missing `private`

File: addons/sys_rack/fnc_unmountRadio.sqf [hunk 14]
```diff
@@ -0,0 +1,37 @@
...
+if (_unit canAdd _mountedRadio) then {
+    [_unit, _mountedRadio] call EFUNC(sys_core,addGear);
+
+    SET_STATE_RACK(_rackId,"mountedRadio","");
+    [_rackId, _mountedRadio] call EFUNC(sys_components,detachAllConnectorsFromComponent);
+
+    //Trigger event
+    [_rackId, "unmountRadio", _mountedRadio] call EFUNC(sys_data,dataEvent);
+} else {
+    systemChat "Unable to unmount radio as you have no inventory space";
```
Comment (Soldia1138) [inline addons/sys_rack/fnc_unmountRadio.sqf]: Should use hint or notification

File: addons/sys_rack/vrc103/__PREP__.sqf [hunk 15]
```diff
@@ -0,0 +1,7 @@
+
+#include "script_component.hpp"
+TRACE_1("enter", _this);
+
+PREP_MODULE(vrc103,mountRadio103);
```
Comment (Soldia1138) [inline addons/sys_rack/vrc103/__PREP__.sqf]: Hmm, not 100% happy with the naming.

IMO it should be RackID first, function second. Like `acre_sys_rack_fnc_vrc103MountRadio`

File: addons/sys_rack/vrc103/fnc_mountableRadio103.sqf [hunk 16]
```diff
@@ -0,0 +1,29 @@
...
+if (([_rackId] call FUNC(getMountedRadio)) != "") exitWith {false}; // If a radio is already mounted we can't mount another.
+
+private _mountable = false;
+private _radioId = _eventData;
+private _baseClass = toLower (BASE_CLASS_CONFIG(_radioId));
+
+if (_baseClass in ["acre_prc117f"]) then { _mountable = true; };
+
+
+_mountable;
```
Comment (Soldia1138) [inline addons/sys_rack/vrc103/fnc_mountableRadio103.sqf]: No New Line at EOF

File: addons/sys_rack/vrc103/script_component.hpp [hunk 17]
```diff
@@ -0,0 +1 @@
+#include "\idi\acre\addons\sys_rack\script_component.hpp"
```
Comment (Soldia1138) [inline addons/sys_rack/vrc103/script_component.hpp]: No New Line at EOF

File: addons/sys_rack/vrc110/__PREP__.sqf [hunk 18]
```diff
@@ -0,0 +1,7 @@
+
+#include "script_component.hpp"
+TRACE_1("enter", _this);
+
+PREP_MODULE(vrc110,mountRadio110);
```
Comment (Soldia1138) [inline addons/sys_rack/vrc110/__PREP__.sqf]: Same as above

File: addons/sys_rack/vrc110/fnc_mountableRadio110.sqf [hunk 19]
```diff
@@ -0,0 +1,28 @@
...
+
+if (([_rackId] call FUNC(getMountedRadio)) != "") exitWith {false}; // If a radio is already mounted we can't mount another.
+
+private _mountable = false;
+private _radioId = _eventData;
+private _baseClass = toLower (BASE_CLASS_CONFIG(_radioId));
+
+if (_baseClass in ["acre_prc152"]) then { _mountable = true; };
+
+_mountable;
```
Comment (Soldia1138) [inline addons/sys_rack/vrc110/fnc_mountableRadio110.sqf]: No New Line at EOF

File: addons/sys_rack/vrc110/script_component.hpp [hunk 20]
```diff
@@ -0,0 +1 @@
+#include "\idi\acre\addons\sys_rack\script_component.hpp"
```
Comment (Soldia1138) [inline addons/sys_rack/vrc110/script_component.hpp]: No New Line at EOF

File: addons/sys_rack/vrc111/__PREP__.sqf [hunk 21]
```diff
@@ -0,0 +1,7 @@
+
+#include "script_component.hpp"
+TRACE_1("enter", _this);
+
+PREP_MODULE(vrc111,mountRadio111);
```
Comment (Soldia1138) [inline addons/sys_rack/vrc111/__PREP__.sqf]: Same as above (naming)

File: addons/sys_rack/vrc111/fnc_mountableRadio111.sqf [hunk 22]
```diff
@@ -0,0 +1,28 @@
...
+
+if (([_rackId] call FUNC(getMountedRadio)) != "") exitWith {false}; // If a radio is already mounted we can't mount another.
+
+private _mountable = false;
+private _radioId = _eventData;
+private _baseClass = toLower (BASE_CLASS_CONFIG(_radioId));
+
+if (_baseClass in ["acre_prc148"]) then { _mountable = true; };
+
+_mountable;
```
Comment (Soldia1138) [inline addons/sys_rack/vrc111/fnc_mountableRadio111.sqf]: No New Line at EOF

File: addons/sys_rack/vrc111/script_component.hpp [hunk 23]
```diff
@@ -0,0 +1 @@
+#include "\idi\acre\addons\sys_rack\script_component.hpp"
```
Comment (Soldia1138) [inline addons/sys_rack/vrc111/script_component.hpp]: No New Line at EOF

**Review by Soldia1138 — COMMENTED** [review]
First run

File: addons/sys_rack/vrc103/fnc_mountableRadio103.sqf (hunk 16 above)
Comment (jonpas) [reply addons/sys_rack/vrc103/fnc_mountableRadio103.sqf]: And no semi-colon for return.

File: addons/sys_rack/fnc_addRackOnReturn.sqf (hunk 9 above)
Comment (jonpas) [reply addons/sys_rack/fnc_addRackOnReturn.sqf:26]: -> #315

**Review by jonpas — APPROVED** [review]

**Pull request merged**